kernel/sched.c

```diff
@@ -794,26 +794,48 @@ ALWAYS_INLINE void z_unpend_thread_no_timeout(struct k_thread *thread)
 	}
 }
 
-#ifdef CONFIG_SYS_CLOCK_EXISTS
-/* Timeout handler for *_thread_timeout() APIs */
-void z_thread_timeout(struct _timeout *timeout)
+/* z_sched_wake_thread is a routine to wake a thread.
+ * is_timeout : If False, clears the prevent_timeout,
+ * If True, prevent_timeout remains unchanged.
+ */
+void z_sched_wake_thread(struct k_thread *thread, bool is_timeout)
 {
-	struct k_thread *thread = CONTAINER_OF(timeout,
-					       struct k_thread, base.timeout);
-
 	LOCKED(&sched_spinlock) {
 		bool killed = ((thread->base.thread_state & _THREAD_DEAD) ||
 			       (thread->base.thread_state & _THREAD_ABORTING));
+		bool prevent;
 
-		if (!killed) {
+#ifdef CONFIG_EVENTS
+		/* Conditionally clear prevent_timeout */
+		thread->prevent_timeout &= is_timeout;
+		prevent = thread->prevent_timeout;
```

Comment:
Having the reset to prevent_timeout happen implicitly seems a little confusing.  Why must this happen here, and only once?  Why not reset it when the next timeout gets set on the thread instead?

Reply:
thread->prevent_timeout value gets cleared up here if this routine is called from the context of k_events_post_internal. Its value is set to true when we are in the middle of processing k_events_post_internal and do not want those threads to timeout and execute in the background.  
So we clear this up here for every thread in waitq which was waiting on events, so we can actually continue with timeouts (line 816) after walking the list of threads is complete. It's happening implicitly here just coz we call this routine from both contexts: 1) An actual timeout and 2) preventing a timeout.